File: day-03/solution.go
package main

import (
	cio "commons"
	"strconv"
	"strings"
)

func part1(lines []string) int64 {
	rays := len(lines[0])
	points := len(lines)

	rayVals := make([]int, rays)

	for _, line := range lines {
		for i, v := range line {
			if v == '1' {
				rayVals[i]++
			}
		}
	}

	gammaRate := make([]string, rays)
	epsilonRate := make([]string, rays)

	for i, val := range rayVals {
		if val > points/2 {
			gammaRate[i] = "1"
		} else {
			gammaRate[i] = "0"
		}

		if val < points/2 {
			epsilonRate[i] = "1"
		} else {
			epsilonRate[i] = "0"
		}
	}

	gammaPower, _ := strconv.ParseInt(strings.Join(gammaRate, ""), 2, 64)
	epsilonPower, _ := strconv.ParseInt(strings.Join(epsilonRate, ""), 2, 64)

	println(gammaPower)
	println(epsilonPower)

	return epsilonPower * gammaPower

}

func main() {
	lines := cio.ReadLines("day-03/input.txt")
	powerRating := part1(lines)

	println(powerRating)

}

File: day-03/solution.ts
import { readLines } from "../commons/io.ts";

function part1(lines: Array<string>): number {
  const rays = lines[0].length;
  const points = lines.length;

  const rayVals = new Array(rays).fill(0);

  lines.forEach((line) => {
    line.split("").forEach((char, index) => {
      if (char === "1") rayVals[index]++;
    });
  });
  const gammaRate = rayVals.map((val) => (val > points / 2) ? 1 : 0);
  const epsilonRate = rayVals.map((val) => (val < points / 2) ? 1 : 0);

  const gammaPower = parseInt(gammaRate.join(""), 2);
  const epsilonPower = parseInt(epsilonRate.join(""), 2);

  console.log(gammaPower);
  console.log(epsilonPower);

  return gammaPower * epsilonPower;
}

function main() {
  const lines = readLines("day-03/input.txt");

  const powerRating = part1(lines);
  console.log(powerRating);
}

main();